src/c++/perf_analyzer/docs/examples/profile.py

```diff
@@ -175,73 +223,134 @@ def collect_periodic_latencies(args, filename):
+    return first_token_latencies, generation_latencies, token_to_token_latencies
+
+
+def calculate_online_metrics(args, profile_result, filename):
+    """Calculate online metrics for more fine-grained performance information."""
+    if not args.stream:
+        return  # skip if offline
+
+    requests = load_json_data(filename)
+    latencies = collect_latencies(requests)
+    first_token_latencies, generation_latencies, token_to_token_latencies = latencies
+
+    profile_result.avg_first_token_latency = np.mean(first_token_latencies)
+    profile_result.avg_total_t2t_latency = np.mean(token_to_token_latencies)
+
+    profile_result.max_first_token_latency = max(first_token_latencies)
+    profile_result.min_first_token_latency = min(first_token_latencies)
+    profile_result.avg_first_token_latency = np.mean(first_token_latencies)
+    profile_result.p50_first_token_latency = np.percentile(
+        first_token_latencies, 50, method="lower"
+    )
+    profile_result.p90_first_token_latency = np.percentile(
+        first_token_latencies, 90, method="lower"
+    )
+    profile_result.p95_first_token_latency = np.percentile(
+        first_token_latencies, 95, method="lower"
+    )
+    profile_result.p99_first_token_latency = np.percentile(
+        first_token_latencies, 99, method="lower"
+    )
+
+    profile_result.max_gen_latency = max(generation_latencies)
+    profile_result.min_gen_latency = min(generation_latencies)
+    profile_result.avg_gen_latency = np.mean(generation_latencies)
+    profile_result.p50_gen_latency = np.percentile(
+        generation_latencies, 50, method="lower"
+    )
+    profile_result.p90_gen_latency = np.percentile(
+        generation_latencies, 90, method="lower"
+    )
+    profile_result.p95_gen_latency = np.percentile(
+        generation_latencies, 95, method="lower"
+    )
+    profile_result.p99_gen_latency = np.percentile(
+        generation_latencies, 99, method="lower"
+    )
+
+    token_latencies = [t / args.max_tokens for t in generation_latencies]
+    profile_result.avg_token_latency = np.mean(token_latencies)
+
+
+def collect_offline_metrics(requests, sequence_len):
+    end_to_end_latencies = []
+    throughputs = []
+    requests = requests["experiments"][0]["requests"]
+
     for request in requests:
-        first_response, *remaining_responses, _ = request["response_timestamps"]
-        first_token_latencies.append(first_response - request["timestamp"])
-        prev_response = first_response
-        for response in remaining_responses:
-            token_to_token_latencies.append(response - prev_response)
-            prev_response = response
-    return first_token_latencies, token_to_token_latencies
+        total_time = request["response_timestamps"][-1] - request["timestamp"]
+        time_s = total_time / 1_000_000_000  # sec
+        time_ms = total_time / 1_000_000  # msec
+        end_to_end_latencies.append(time_ms)
+        throughputs.append(sequence_len / time_s)
```

Comment:
Should we distinguish between throughputs that include the prefill (prompt) and throughputs that don't? This includes the prefill, but do we think users might want to know "output-only" throughput? Similar to how the vLLM benchmarking script provides average per-token latency and average per-output-token latency?

Reply:
Added generation throughput metrics to the results.